contracts/data-storage/andromeda-counter/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, ensure, Storage, Addr};
+
+use andromeda_data_storage::counter::{ExecuteMsg, InstantiateMsg, QueryMsg, CounterRestriction};
+use andromeda_data_storage::counter::{
+    GetInitialAmountResponse, GetCurrentAmountResponse, GetIncreaseAmountResponse, GetDecreaseAmountResponse, GetRestrictionResponse,
+};
+use andromeda_std::{
+    ado_base::{InstantiateMsg as BaseInstantiateMsg, MigrateMsg},
+    ado_contract::ADOContract,
+    common::{context::ExecuteContext, encode_binary},
+    error::ContractError,
+};
+use cw_utils::nonpayable;
+
+use crate::state::{RESTRICTION, CURRENT_AMOUNT, DEFAULT_INITIAL_AMOUNT, DEFAULT_INCREASE_AMOUNT, DEFAULT_DECREASE_AMOUNT, INITIAL_AMOUNT, INCREASE_AMOUNT, DECREASE_AMOUNT};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:andromeda-counter";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let resp = ADOContract::default().instantiate(
+        deps.storage,
+        env,
+        deps.api,
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+
+    RESTRICTION.save(deps.storage, &msg.restriction)?;
+
+    if let Some(initial_amount) = msg.initial_amount {
```

Comment:
These lines could be simplified by directly saving `msg.initial_amount.unwrap_or(default_amount)`